cirq-core/cirq/sim/simulation_state_base.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An interface for quantum states as targets for operations."""
+import abc
+from typing import (
+    Any,
+    Dict,
+    Generic,
+    Iterator,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    TYPE_CHECKING,
+    TypeVar,
+    Union,
+)
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    import cirq
+
+
+TSelfTarget = TypeVar('TSelfTarget', bound='SimulationStateBase')
+TSimulationState = TypeVar('TSimulationState', bound='cirq.SimulationState')
+
+
+class SimulationStateBase(Generic[TSimulationState], metaclass=abc.ABCMeta):
```

Comment:
@daxfohl Can you please explain what's going on with the generics here? For example:

1) `SimulationStateBase(Generic[TSimulationState], metaclass=abc.ABCMeta)` is a generic of type `TSimulationState` = TypeVar('TSimulationState', bound='cirq.SimulationState')` and

2) `SimulationState(SimulationStateBase, Generic[TState], metaclass=abc.ABCMeta)` derives from `SimulationStateBase` and is also a generic of type `TypeVar('TState', bound='cirq.QuantumStateRepresentation')`

Why is `TSimulationState` used in `SimulatorStateBase` bound to `SimulatorState` and not `SimulatorStateBase` ? Why can we not use `TSelfTarget ` in place of `TSimulationState` in the `SimulationStateBase` class? Is it because we want to allow `__getitem__` and `create_merged_state` to potentially return a different type compared to `TSelfTarget` ?

Reply:
Thanks for asking. It's pretty challenging and I remember some idiosyncrasies of mypy getting in the way for a while.

So hierarchy: SimStateBase has two subclasses {SimState, SimProductState}. 

SimState is abstract and has multiple subclasses: StateVectorSimState, DensityMatrixSimState, etc. Those are "dense" as in they have a single big state vector, density matrix, tableau, or whatever for all qubits in the simulation. TSimState is the generic constrained to a single SimState.

SimProductState is concrete and generic: SimProductState[TSimState]. It contains a qubit->TSimState mapping. It's a product state representation for whichever TSimState it contains, so it's smaller and faster when there's not full entanglement. I consider this a "final" class. (You can control whether a simulator uses a SimProductState or a single SimState, by the `split_untangled_states` flag in simulator constructors that support it).

SimStateBase is the superclass of them all. It is generic on TSimState, because there's an abstract function `create_merged_state() -> TSimState` that's there to give you the "dense" version of whatever your TSimState is. So in SimState, the implementation of that just returns self; in SimProductState it krons up all the values in its mapping and returns that. This is why SimStateBase is generic on TSimState, and why TSimState is bound to SimState: SimState is the "dense" representation. `__getitem__` works similarly; the docstring should probably read "Gets the dense substate associated with the qubit" or something like that.

Then TSelfTarget is different because the `copy` of a SimProductState should return a SimProductState.

Reply:
And `TState` is separate. It is purely the quantum component. SimState contains not only the quantum component (density matrix, tableau, etc), but also the qubit->axis mapping, a prng, measurement records, etc. So that's why you see DensityMatrixSimState inherits SimState[DensityMatrixQuantumRepresentation].

In an ideal world we shouldn't even need DensityMatrixSimState etc as their own classes, and instead just have SimState be a generic final class that we use directly. However there are a lot of backwards compatibility issues and useful convenience methods preventing that. (Otherwise you'd end up needing to write code like `result.get_merged_state().state.state.state_vector` or something).

Reply:
Thanks, this is extremely helpful! 

I think the fact that we are supporting `SimulationProductState` for any arbitrary `TSimulationState` is adding a lot of complexity to the type hierarchy. An alternative could have been to have 
1) A type Hierarchy for specifying simulation states as `Generic[TState]`; i.e. what the `SimulationState` class is doing right now. `SimulationStateBase` wouldn't exist and the abstract base class for simulation states would be `SimulationState`. 
2) Have a separate container type for dense vs sparse representation of states -- This container could maintain the `Dict[cirq.Qid, TSimulationState]`. Based on the configurable parameter, a simulator would construct either a dense container or a sparse container for the specified simulation state type. 

The fact that a container of simulation states (`SimulationProductState`) is also a simulation state is I think making the type hierarchy super complex. But given the Cirq 1.0 backwards compatibility constraints, I'm not sure if we can do anything about the situation at this point except improving documentation.

Reply:
Yes, that was our initial approach but that caused some backwards incompatibility at the time IIRC, so we went with subclassing. Probably could have revisited this before 1.0 but didn't think about it.